crates/ra_cli/src/analysis_stats.rs

```diff
@@ -100,6 +101,40 @@ pub fn run(verbose: bool, memory_usage: bool, path: &Path, only: Option<&str>) -
                     num_exprs_partially_unknown += 1;
                 }
             }
+            if let Some(mismatch) = inference_result.type_mismatch_for_expr(expr_id) {
+                num_type_mismatches += 1;
+                if verbose {
+                    let src = f.source(db);
+                    let original_file = src.file_id.original_file(db);
+                    let path = db.file_relative_path(original_file);
+                    let line_index = host.analysis().file_line_index(original_file).unwrap();
+                    let body_source_map = f.body_source_map(db);
+                    let syntax_node = body_source_map.expr_syntax(expr_id);
+                    let line_col = syntax_node.map(|syntax_node| {
+                        (
+                            line_index.line_col(syntax_node.range().start()),
+                            line_index.line_col(syntax_node.range().end()),
+                        )
+                    });
+                    let line_col = match line_col {
```

Comment:
Yeah, I feel like this is roughtly the correct way to get location. 

We try very hard to make hir location-less, which will also helps quite a bunch when we supplement our "from source" analysis with "from rlib" analysis, where locations could be optional. 

Maaaybe we should make it possible to expose this information convenintly via `fmt::Debug`, but wouldn't want to add this a first-class functions to hir

Reply:
although it sems like we perhaps miss `.source(ExprId) -> Source<ast::Expr>` method on `SourceAnalyzer`?

Reply:
That sounds like an improvement already!

Reply:
Hmm. This doesn't work so easily, because `ExprId`s don't necessarily point to `ast::Expr`s, but also to `ast::Block`s... so we'd need a `ExprOrBlock` or something like that, or we could return `None` in that case (with the reasoning that the `Block` basically gets desugared to a `BlockExpr`, which doesn't exist in the actual source)...

Reply:
I suggest we just return `None` for the time being.

In general, our ast_source_map is pretty wonky and lossy, especially around macros/deshugaring. I'd like to fix it in a separate bit of work. And this actually also another reason to merge `ast::Block` and `ast::BlockExpr` !

Reply:
Ok, that's what I did for now.